### PR TITLE
port PeerFsm step 2

### PR DESCRIPTION
**Pull request by coocood**
Refactor `Msg` to use `interface` instead of multiple fields.
Since the struct is small, only 32 bytes, we can use value instead of pointer.
Move some code from peer to fsm_apply and fsm_store.

### CODE REVIEW

File: tikv/raftstore/fsm_peer.go [hunk 1]
```diff
@@ -464,19 +474,165 @@ func (d *peerFsmDelegate) validateRaftMessage(msg *rspb.RaftMessage) bool {
...
 }
 
 func (d *peerFsmDelegate) needGCMerge(msg *rspb.RaftMessage) (bool, error) {
 	return false, nil // TODO: merge func
 }
 
 func (d *peerFsmDelegate) handleGCPeerMsg(msg *rspb.RaftMessage) {
-	// TODO: stub
+	fromEpoch := msg.RegionEpoch
+	if !IsEpochStale(fromEpoch, d.peer.Region().RegionEpoch) {
```
Comment (zhangjinpeng1987) [inline tikv/raftstore/fsm_peer.go]: Seems `fromEpoch` should be the second parameter?

File: tikv/raftstore/fsm_peer.go [hunk 2]
```diff
@@ -464,19 +474,165 @@ func (d *peerFsmDelegate) validateRaftMessage(msg *rspb.RaftMessage) bool {
...
+	meta := d.ctx.storeMeta
+	if !RegionEqual(meta.regions[d.regionID()], d.region()) {
+		if !d.peer.isInitialized() {
+			log.Infof("%s stale delegate detected, skip", d.tag())
+			return &key, nil
+		} else {
+			panic(fmt.Sprintf("%s meta corrupted %s != %s", d.tag(), meta.regions[d.regionID()], d.region()))
+		}
+	}
+	existOverlapRegion := d.findOverlapRegion(meta, snapRegion)
```
Comment (zhangjinpeng1987) [inline tikv/raftstore/fsm_peer.go]: I don't see this logic in TiKV repo.

File: tikv/raftstore/fsm_store.go [hunk 3]
```diff
@@ -1,5 +1,88 @@
...
+		FromPeer:    ClonePeer(fromPeer),
+		ToPeer:      ClonePeer(toPeer),
+		RegionEpoch: CloneRegionEpoch(curEpoch),
+	}
+	if targetRegion != nil {
+		gcMsg.MergeTarget = targetRegion
+	} else {
+		gcMsg.IsTombstone = true
+	}
+	err := pc.trans.Send(gcMsg)
```
Comment (zhangjinpeng1987) [inline tikv/raftstore/fsm_store.go]: if err:= xxx; err != nil {
}

File: tikv/raftstore/fsm_peer.go (hunk 2 above)
Comment (coocood) [reply tikv/raftstore/fsm_peer.go]: This is the corresponding code in tikv
```
        let r = meta
            .region_ranges
            .range((Excluded(enc_start_key(&snap_region)), Unbounded::<Vec<u8>>))
            .map(|(_, &region_id)| &meta.regions[&region_id])
            .take_while(|r| enc_start_key(r) < enc_end_key(&snap_region))
            .skip_while(|r| r.get_id() == region_id)
            .next()
            .map(|r| r.to_owned());
```

**Comment by zhangjinpeng1987** [comment]
LGTM

**Pull request merged**